File: app/src/main/java/com/reteno/sample/fragments/database/UserReadDialogFragment.java
package com.reteno.sample.fragments.database;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.transition.AutoTransition;
import androidx.transition.TransitionManager;

import com.reteno.core.data.local.database.manager.RetenoDatabaseManagerUser;
import com.reteno.core.data.local.model.user.AddressDb;
import com.reteno.core.data.local.model.user.UserAttributesDb;
import com.reteno.core.data.local.model.user.UserCustomFieldDb;
import com.reteno.core.data.local.model.user.UserDb;
import com.reteno.core.data.remote.mapper.JsonMappersKt;
import com.reteno.sample.R;
import com.reteno.sample.databinding.ItemDbUserBinding;
import com.reteno.sample.databinding.ViewUserCustomFieldsVerticalBinding;

import java.util.List;
import java.util.Locale;


public class UserReadDialogFragment extends BaseReadDialogFragment<UserDb, ItemDbUserBinding, UserReadDialogFragment.UserViewHolder, UserReadDialogFragment.UserAdapter> {

    private RetenoDatabaseManagerUser databaseManager;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        databaseManager = serviceLocator.getRetenoDatabaseManagerUserProvider().get();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        databaseManager = null;
    }

    @Override
    protected void initAdapter() {
        adapter = new UserAdapter(() ->
                TransitionManager.beginDelayedTransition(bindingMain.rvList, new AutoTransition())
        );
    }

    @Override
    protected void initCount() {
        long userEventsCount = databaseManager.getUserCount();
        bindingMain.tvCount.setText(String.format(Locale.US, "Count: %d", userEventsCount));
    }

    @Override
    protected void initItems() {
        List<UserDb> newItems = databaseManager.getUsers(null);
        adapter.setItems(newItems);
    }

    @Override
    protected void deleteItems(int count) {
        List<UserDb> users = databaseManager.getUsers(count);
        for (UserDb user : users) {
            databaseManager.deleteUser(user);
        }
    }

    //==============================================================================================
    static class UserAdapter extends BaseReadAdapter<UserDb, ItemDbUserBinding, UserViewHolder> {

        UserAdapter(ViewHolderListener listener) {
            super(listener);
        }

        @NonNull
        @Override
        public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            ItemDbUserBinding binding = ItemDbUserBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new UserViewHolder(binding);
        }

        @Override
        protected void initListeners(ItemDbUserBinding binding) {
            binding.ivExpand.setOnClickListener(v -> {
                if (binding.llContent.getVisibility() == View.VISIBLE) {
                    onExpandCollapseClickListener.onExpandCollapse();
                    binding.llContent.setVisibility(View.GONE);
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_more);
                } else {
                    onExpandCollapseClickListener.onExpandCollapse();
                    binding.llContent.setVisibility(View.VISIBLE);
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_less);
                }
            });
        }
    }

    //==============================================================================================
    static class UserViewHolder extends BaseReadViewHolder<UserDb, ItemDbUserBinding> {

        UserViewHolder(ItemDbUserBinding binding) {
            super(binding);
        }

        @Override
        protected void bind(UserDb user) {
            bindingHolder.tvDeviceId.setTextOrHide(user.getDeviceId());
            bindingHolder.tvExternalUserId.setTextOrHide(user.getExternalUserId());
            bindingHolder.tvSubscriptionKeys.setTextOrHide(JsonMappersKt.toJsonOrNull(user.getSubscriptionKeys()));
            bindingHolder.tvGroupNamesInclude.setTextOrHide(JsonMappersKt.toJsonOrNull(user.getGroupNamesInclude()));
            bindingHolder.tvGroupNamesInclude.setTextOrHide(JsonMappersKt.toJsonOrNull(user.getGroupNamesExclude()));

            UserAttributesDb userAttributes = user.getUserAttributes();
            if (userAttributes != null) {
                bindingHolder.llUserAttributes.setVisibility(View.VISIBLE);
                bindingHolder.tvPhone.setTextOrHide(userAttributes.getPhone());
                bindingHolder.tvEmail.setTextOrHide(userAttributes.getEmail());
                bindingHolder.tvFirstName.setTextOrHide(userAttributes.getFirstName());
                bindingHolder.tvLastName.setTextOrHide(userAttributes.getLastName());
                bindingHolder.tvLanguageCode.setTextOrHide(userAttributes.getLanguageCode());
                bindingHolder.tvTimeZone.setTextOrHide(userAttributes.getTimeZone());

                AddressDb address = userAttributes.getAddress();
                if (address != null) {
                    bindingHolder.llAddress.setVisibility(View.VISIBLE);
                    bindingHolder.tvRegion.setTextOrHide(address.getRegion());
                    bindingHolder.tvTown.setTextOrHide(address.getTown());
                    bindingHolder.tvAddress.setTextOrHide(address.getAddress());
                    bindingHolder.tvPostCode.setTextOrHide(address.getPostcode());
                } else {
                    bindingHolder.llAddress.setVisibility(View.GONE);
                }

                List<UserCustomFieldDb> customFields = userAttributes.getFields();
                if (customFields != null && !customFields.isEmpty()) {
                    bindingHolder.llCustomFields.setVisibility(View.VISIBLE);
                    for (UserCustomFieldDb customField : customFields) {
                        View customFieldView = createNewFields(customField.getKey(), customField.getValue());
                        bindingHolder.llCustomFields.addView(customFieldView);
                    }
                } else {
                    bindingHolder.llCustomFields.setVisibility(View.GONE);
                }
            } else {
                bindingHolder.llUserAttributes.setVisibility(View.GONE);
            }
        }

        private View createNewFields(String key, String value) {
            ViewUserCustomFieldsVerticalBinding binding = ViewUserCustomFieldsVerticalBinding.inflate(LayoutInflater
                    .from(bindingHolder.getRoot().getContext()), bindingHolder.llCustomFields, false);
            binding.etKey.setText(key);
            binding.etValue.setText(value);
            return binding.getRoot();
        }
    }
}

File: app/src/main/java/com/reteno/sample/fragments/database/UserReadDialogFragment.kt
package com.reteno.sample.fragments.database

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.transition.AutoTransition
import androidx.transition.TransitionManager
import com.reteno.core.data.local.database.manager.RetenoDatabaseManagerUser
import com.reteno.core.data.local.model.user.UserDb
import com.reteno.core.data.remote.mapper.toJsonOrNull
import com.reteno.sample.R
import com.reteno.sample.databinding.ItemDbUserBinding
import com.reteno.sample.databinding.ViewUserCustomFieldsVerticalBinding
import com.reteno.sample.fragments.database.UserReadDialogFragment.UserAdapter
import com.reteno.sample.fragments.database.UserReadDialogFragment.UserViewHolder
import java.util.Locale

internal class UserReadDialogFragment :
    BaseReadDialogFragment<UserDb, ItemDbUserBinding, UserViewHolder, UserAdapter>() {
    private var databaseManager: RetenoDatabaseManagerUser? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        databaseManager = serviceLocator.retenoDatabaseManagerUserProvider.get()
    }

    override fun onDestroy() {
        super.onDestroy()
        databaseManager = null
    }

    override fun initAdapter() {
        adapter = UserAdapter {
            TransitionManager.beginDelayedTransition(
                bindingMain!!.rvList, AutoTransition()
            )
        }
    }

    override fun initCount() {
        val userEventsCount = databaseManager!!.getUnSyncedUserCount()
        bindingMain!!.tvCount.text = String.format(Locale.US, "Count: %d", userEventsCount)
    }

    override fun initItems() {
        val newItems = databaseManager!!.getUsers(null)
        adapter!!.setItems(newItems)
    }

    override fun deleteItems(count: Int) {
        val users = databaseManager!!.getUsers(count)
        for (user in users) {
            databaseManager!!.deleteUser(user)
        }
    }

    //==============================================================================================
    internal class UserAdapter(listener: ViewHolderListener) :
        BaseReadAdapter<UserDb, ItemDbUserBinding, UserViewHolder>(listener) {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
            val binding =
                ItemDbUserBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return UserViewHolder(binding)
        }

        override fun initListeners(binding: ItemDbUserBinding) {
            binding.ivExpand.setOnClickListener {
                if (binding.llContent.visibility == View.VISIBLE) {
                    onExpandCollapseClickListener.onExpandCollapse()
                    binding.llContent.visibility = View.GONE
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_more)
                } else {
                    onExpandCollapseClickListener.onExpandCollapse()
                    binding.llContent.visibility = View.VISIBLE
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_less)
                }
            }
        }
    }

    //==============================================================================================
    class UserViewHolder(binding: ItemDbUserBinding) :
        BaseReadViewHolder<UserDb, ItemDbUserBinding>(binding) {
        override fun bind(user: UserDb) {
            bindingHolder.tvDeviceId.setTextOrHide(user.deviceId)
            bindingHolder.tvExternalUserId.setTextOrHide(user.externalUserId)
            bindingHolder.tvSubscriptionKeys.setTextOrHide(user.subscriptionKeys.toJsonOrNull())
            bindingHolder.tvGroupNamesInclude.setTextOrHide(user.groupNamesInclude.toJsonOrNull())
            bindingHolder.tvGroupNamesInclude.setTextOrHide(user.groupNamesExclude.toJsonOrNull())
            val userAttributes = user.userAttributes
            if (userAttributes != null) {
                bindingHolder.llUserAttributes.visibility = View.VISIBLE
                bindingHolder.tvPhone.setTextOrHide(userAttributes.phone)
                bindingHolder.tvEmail.setTextOrHide(userAttributes.email)
                bindingHolder.tvFirstName.setTextOrHide(userAttributes.firstName)
                bindingHolder.tvLastName.setTextOrHide(userAttributes.lastName)
                bindingHolder.tvLanguageCode.setTextOrHide(userAttributes.languageCode)
                bindingHolder.tvTimeZone.setTextOrHide(userAttributes.timeZone)
                val address = userAttributes.address
                if (address != null) {
                    bindingHolder.llAddress.visibility = View.VISIBLE
                    bindingHolder.tvRegion.setTextOrHide(address.region)
                    bindingHolder.tvTown.setTextOrHide(address.town)
                    bindingHolder.tvAddress.setTextOrHide(address.address)
                    bindingHolder.tvPostCode.setTextOrHide(address.postcode)
                } else {
                    bindingHolder.llAddress.visibility = View.GONE
                }
                val customFields = userAttributes.fields
                if (!customFields.isNullOrEmpty()) {
                    bindingHolder.llCustomFields.visibility = View.VISIBLE
                    for ((key, value) in customFields) {
                        val customFieldView = createNewFields(key, value)
                        bindingHolder.llCustomFields.addView(customFieldView)
                    }
                } else {
                    bindingHolder.llCustomFields.visibility = View.GONE
                }
            } else {
                bindingHolder.llUserAttributes.visibility = View.GONE
            }
        }

        private fun createNewFields(key: String, value: String?): View {
            val binding = ViewUserCustomFieldsVerticalBinding.inflate(
                LayoutInflater
                    .from(bindingHolder.getRoot().context), bindingHolder.llCustomFields, false
            )
            binding.etKey.setText(key)
            binding.etValue.setText(value)
            return binding.getRoot()
        }
    }
}
